### PR TITLE
nix-serve: fix NIX_SECRET_KEY_FILE

### PR DESCRIPTION
**Pull request by Artturin**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/nix-serve.nix [hunk 1]
```diff
@@ -61,16 +61,19 @@ in
...
 
       serviceConfig = {
         Restart = "always";
         RestartSec = "5s";
-        ExecStart = "${pkgs.nix-serve}/bin/nix-serve " +
+        # does not work due to the cred var not being substituted
+        # environment.NIX_SECRET_KEY_FILE = "$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE";
+        ExecStart = "NIX_SECRET_KEY_FILE=$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE " +
+          "${pkgs.nix-serve}/bin/nix-serve " +
           "--listen ${cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}";
```
Comment (pennae) [inline nixos/modules/services/networking/nix-serve.nix:72]:
```suggestion
        script = ''
          export NIX_SECRET_KEY_FILE="$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE"
          exec ${pkgs.nix-serve}/bin/nix-serve \
            --listen ${escapeShellArg cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}
        '';
```

File: nixos/modules/services/networking/nix-serve.nix [hunk 2]
```diff
@@ -61,16 +61,19 @@ in
 
       path = [ config.nix.package.out pkgs.bzip2.bin ];
       environment.NIX_REMOTE = "daemon";
-      environment.NIX_SECRET_KEY_FILE = cfg.secretKeyFile;
+
+      script = ''
+        export NIX_SECRET_KEY_FILE="$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE"
+        ${pkgs.nix-serve}/bin/nix-serve --listen ${cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}
```
Comment (alexbakker) [inline nixos/modules/services/networking/nix-serve.nix:67]: Should probably prepend ``exec`` here.

**Comment by alexbakker** [comment]
I left one nitpick comment.

Other than that: I tested this on 21.11 and it works well.

**Review by alexbakker — APPROVED** [review]

**Comment by Mic92** [comment]
@ofborg test nix-serve

File: nixos/modules/services/networking/nix-serve.nix [hunk 3]
```diff
@@ -61,16 +59,19 @@ in
...
 
       serviceConfig = {
         Restart = "always";
         RestartSec = "5s";
-        ExecStart = "${pkgs.nix-serve}/bin/nix-serve " +
-          "--listen ${cfg.bindAddress}:${toString cfg.port} ${cfg.extraParams}";
         User = "nix-serve";
         Group = "nix-serve";
         DynamicUser = true;
+        LoadCredential = "NIX_SECRET_KEY_FILE:${cfg.secretKeyFile}";
```
Comment (Mic92) [inline nixos/modules/services/networking/nix-serve.nix:74]:
```suggestion
        LoadCredential = lib.optionalString (cfg.secretKeyFile != null) "NIX_SECRET_KEY_FILE:${cfg.secretKeyFile}";
```

File: nixos/modules/services/networking/nix-serve.nix [hunk 4]
```diff
@@ -61,16 +59,19 @@ in
 
       path = [ config.nix.package.out pkgs.bzip2.bin ];
       environment.NIX_REMOTE = "daemon";
-      environment.NIX_SECRET_KEY_FILE = cfg.secretKeyFile;
+
+      script = ''
+        export NIX_SECRET_KEY_FILE="$CREDENTIALS_DIRECTORY/NIX_SECRET_KEY_FILE"
```
Comment (Mic92) [inline nixos/modules/services/networking/nix-serve.nix:64]: This also might have to be wrapped in an optional string than if we have no secret file.

File: nixos/modules/services/networking/nix-serve.nix (hunk 3 above)
Comment (Artturin) [reply nixos/modules/services/networking/nix-serve.nix]: thanks

**Comment by Artturin** [comment]
the test fails with `vm-test-run-nix-ssh-serve> client # error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`

**Comment by Mic92** [comment]
> the test fails with `vm-test-run-nix-ssh-serve> client # error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`

I cannot reproduce this error locally, can you?

**Comment by Artturin** [comment]
> > the test fails with `vm-test-run-nix-ssh-serve> client # error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`
> 
> I cannot reproduce this error locally, can you?

yes at https://github.com/nixos/nixpkgs/commit/a78dd785b29bbbcad5490282308b9dbdc244a0f3

**Comment by Mic92** [comment]
It's not even clear to me why there is `nix copy` used.

**Comment by Artturin** [comment]
there's 3 nix commands run in nix-serve https://github.com/edolstra/nix-serve/blob/fefa88d82ead34d1ad97cc68cccae9412accfc1f/nix-serve.psgi

2 `nix dump-path`
1 `nix-log`

so we should pin it to `nix_2_3` for now 
also there's https://github.com/edolstra/nix-serve/issues/28

**Comment by Artturin** [comment]
the nix-serve test is defined wrong in all-tests.nix

```
  nix-serve = handleTest ./nix-ssh-serve.nix {};
  nix-ssh-serve = handleTest ./nix-ssh-serve.nix {};
```
nix-ssh-serve has nix copy

**Comment by Artturin** [comment]
you forgot to change the path after you pasted in ac7b8724b59974c0d74f2feacc4a2a787a5cf122

**Comment by Artturin** [comment]
nevermind no need to pin nix https://github.com/NixOS/nix/pull/5634 this pr fixed the issue

**Pull request merged**